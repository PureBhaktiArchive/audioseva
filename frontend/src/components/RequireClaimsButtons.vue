<template>
  <span>
    <v-btn :key="`${button.text}-${index}`" v-for="(button, index) in routeButtons" v-bind="button.props">
      {{ button.text }}
    </v-btn>
  </span>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import _ from "lodash";
import { mapActions } from "vuex";
import { RouteConfig } from "vue-router";
import { filterRoutesByClaims, hasClaim } from "@/router";

@Component({
  name: "RequireClaimsButtons",
  methods: {
    ...mapActions("user", ["getUserClaims"])
  }
})
export default class RequireClaimsButtons extends Vue {
  userClaims: any = {};
  routeButtons: any[] = [];

  async mounted() {
    this.userClaims = (await this.getUserClaims()) || {};
    this.makeButtons();
  }

  makeButtons() {
    const buttons: any[] = [];
    const defaultButtonProps = {
      color: "primary"
    };
    const filterCb = (
      route: RouteConfig,
      userClaims: any,
      requiredClaims: any,
      parentPath: string
    ): any => {
      const homePageButton = _.get(route, "meta.homePageLink", false);
      const fullPath = `${parentPath}${route.path}`;
      let buttonProps: any;
      if (homePageButton) {
        buttonProps = {
          text: homePageButton.text,
          props: {
            ...defaultButtonProps,
            ...homePageButton.props,
            to: fullPath
          }
        };
      }
      if (requiredClaims) {
        if (hasClaim(requiredClaims, userClaims) && homePageButton) {
          buttons.push(buttonProps);
        }
      } else if (homePageButton) {
        buttons.push(buttonProps);
      }
      if (route.children) {
        filterRoutesByClaims(filterCb)(
          route.children,
          userClaims,
          requiredClaims,
          fullPath
        );
      }
    };

    const routes = (this.$router as any).options.routes.find(
      (route: any) => route.path === "/"
    ).children;
    filterRoutesByClaims(filterCb)(routes, this.userClaims, false, "/");
    this.routeButtons = buttons;
  }
}
</script>

<style scoped>
</style>
