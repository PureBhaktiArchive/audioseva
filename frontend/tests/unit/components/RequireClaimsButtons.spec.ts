import Vue from "vue";
import { mount, createLocalVue } from "@vue/test-utils";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import RequireClaimsButtons from "@/components/RequireClaimsButtons.vue";

const router = new VueRouter({
  mode: "history",
  routes: [
    {
      path: "/",
      meta: { auth: { requireAuth: true } },
      component: () => import("@/views/Dashboard.vue"),
      children: [
        {
          path: "",
          component: () => import("@/views/Home.vue")
        },
        {
          path: "users",
          component: () => import("@/views/Users/List.vue"),
          meta: { menuItem: true, menuName: "People", menuIcon: "fas fa-users" }
        },
        {
          path: "sqr/",
          component: () => import("@/views/SQR/SQR.vue"),
          meta: {
            activator: true,
            activatorName: "Sound Quality Reporting",
            menuIcon: "fas fa-headphones",
            auth: { requireClaims: { SQR: true } }
          },
          children: [
            {
              path: "",
              component: () => import("@/views/SQR/Files.vue"),
              meta: {
                menuItem: true,
                homePageLink: { text: "SQR" }
              }
            },
            {
              path: "allot",
              component: () => import("@/views/SQRAllotment.vue"),
              meta: { menuItem: true }
            }
          ]
        },
        {
          path: "te/",
          component: () => import("@/views/TE/TE.vue"),
          meta: {
            activator: true,
            activatorName: "Track Editing",
            menuIcon: "fas fa-cut"
          },
          children: [
            {
              path: "tasks",
              component: () => import("@/views/TE/Tasks.vue"),
              meta: {
                menuItem: true,
                homePageLink: { text: "Tasks" }
              }
            },
            {
              path: "allot",
              component: () => import("@/views/TE/Allotment.vue"),
              meta: { menuItem: true }
            },
            {
              path: "my",
              component: () => import("@/views/TE/MyTasks.vue"),
              meta: { menuItem: true, menuLinkName: "My Tasks" }
            },
            {
              path: "upload",
              component: () => import("@/views/TE/Upload.vue"),
              meta: {
                menuItem: true,
                auth: { requireClaims: { TE: true } },
                homePageLink: { text: "Upload" }
              }
            },
            {
              path: "tasks/:taskId",
              component: () => import("@/views/TE/Task.vue"),
              meta: {
                menuItem: false,
                requireClaims: { TE: true, coordinator: true }
              }
            }
          ]
        }
      ]
    }
  ]
});

Vue.use(Vuetify);

describe("RequireClaimsButtons", () => {
  it("should render buttons that match user claims", () => {
    const localVue = createLocalVue();
    localVue.use(VueRouter);
    const wrapper = mount(RequireClaimsButtons, {
      localVue,
      router,
      data: () => ({
        userClaims: { TE: true }
      }),
      methods: {
        getUserClaims: () => {}
      }
    });
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.routeButtons).toMatchSnapshot();
    });
  });
});
