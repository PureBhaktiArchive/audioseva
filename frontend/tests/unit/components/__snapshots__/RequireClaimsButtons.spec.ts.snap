// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequireClaimsButtons should render buttons that match user claims 1`] = `
Array [
  Object {
    "props": Object {
      "color": "primary",
      "to": "/te/tasks",
    },
    "text": "Tasks",
  },
]
`;
